samples/nrf9160/gps/src/main.c

```diff
@@ -128,324 +136,268 @@ static int activate_lte(bool activate)
 
 	return 0;
 }
+
+static void get_agps_data(struct k_work *item)
+{
+	ARG_UNUSED(item);
+
+	printk("\033[1;1H");
+	printk("\033[2J");
+	printk("New A-GPS data requested, contacting SUPL server, flags %d\n",
+			last_agps.data_flags);
+
+	activate_lte(true);
+	printk("Established LTE link\n");
+	if (open_supl_socket() == 0) {
+		printk("Starting SUPL session\n");
+		supl_session(&last_agps);
+		printk("Done\n");
+		close_supl_socket();
+	}
+	activate_lte(false);
+}
+
+static int inject_agps_type(void *agps, size_t agps_size, uint16_t type, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	int retval = nrf_modem_gnss_agps_write(agps, agps_size, type);
+
+	if (retval != 0) {
+		printk("Failed to write A-GPS data, type: %d (errno: %d)\n", type, errno);
+		return -1;
+	}
+
+	printk("Injected A-GPS data, type: %d, size: %d\n", type, agps_size);
+
+	return 0;
+}
 #endif
 
-static int gnss_ctrl(uint32_t ctrl)
+static void gnss_event_handler(int event)
 {
 	int retval;
+	struct nrf_modem_gnss_nmea_data_frame *nmea_data;
 
-	nrf_gnss_fix_retry_t    fix_retry    = 0;
-	nrf_gnss_fix_interval_t fix_interval = 1;
-	nrf_gnss_delete_mask_t	delete_mask  = 0;
-	nrf_gnss_nmea_mask_t	nmea_mask    = NRF_GNSS_NMEA_GSV_MASK |
-					       NRF_GNSS_NMEA_GSA_MASK |
-					       NRF_GNSS_NMEA_GLL_MASK |
-					       NRF_GNSS_NMEA_GGA_MASK |
-					       NRF_GNSS_NMEA_RMC_MASK;
-
-	if (ctrl == GNSS_INIT_AND_START) {
-		gnss_fd = nrf_socket(NRF_AF_LOCAL,
-				     NRF_SOCK_DGRAM,
-				     NRF_PROTO_GNSS);
-
-		if (gnss_fd >= 0) {
-			printk("GPS Socket created\n");
-		} else {
-			printk("Could not init socket (err: %d)\n", gnss_fd);
-			return -1;
-		}
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		nrf_modem_gnss_read(&last_pvt, sizeof(last_pvt), NRF_MODEM_GNSS_DATA_PVT);
+		k_sem_give(&pvt_data_sem);
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_FIX_RETRY,
-					&fix_retry,
-					sizeof(fix_retry));
-		if (retval != 0) {
-			printk("Failed to set fix retry value\n");
-			return -1;
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		nmea_data = k_malloc(sizeof(struct nrf_modem_gnss_nmea_data_frame));
+		if (nmea_data == NULL) {
+			break;
```

Comment:
Would be usueful with a print here that memory allocation failed, indicating memory leak

Reply:
Added the error print, although it is immediately overwritten because the sample clears the screen when it starts printing new PVT data. Anyway, it will still be visible in a log.